torch/_export/utils.py

```diff
@@ -399,3 +409,161 @@ def node_inline_(call_mod_node: torch.fx.Node) -> None:
+    """
+    When torch.cond ops are present, GraphModules contain sub-GraphModules,
+    whose inputs are nodes from the top-level graph.
+    However, the default placeholder node names in lower-level graphs have no relation to their parents.
+    This function propagates the placeholder names from the top-level graph to the sub-graphs.
+    - For params/buffers/constants/user inputs, the same name is kept at all levels.
+    - For call_function inputs, names may be duplicated,
+      (e.g. a subgraph takes a mul node as input, and also performs as mul op)
+      so we use nn_module_stack to describe where the node is from.
+    """
+    for node in gm.graph.nodes:
+        if (
+            node.op == "call_function"
+            and node.name.startswith("conditional")
+            and isinstance(node.target, torch._ops.HigherOrderOperator)
+        ):
+            _, true_getattr, false_getattr, cond_args = node.args
+            # propagate names to both subgraphs
+            for getattr_node in [true_getattr, false_getattr]:
+                subgraph = getattr(gm, getattr_node.name)
+                for i, _node in enumerate(subgraph.graph.nodes):
+                    # placeholder nodes come first
+                    if i >= len(cond_args):
+                        break
+                    # handle duplicate names on ops with nn_module_stack
+                    if cond_args[i].op == "call_function":
+                        module_stack = cond_args[i].meta.get("nn_module_stack", {})
+                        module_stack = [x[0] for x in module_stack.values()]
+                        name = "_".join(
+                            ["_root"] + module_stack[1:] + [cond_args[i].name]
+                        )
+                    else:  # keep same name
+                        name = cond_args[i].name
+                    _node.name = _node.target = name
+                propagate_placeholder_names_for_cond(subgraph)  # recurse
+                subgraph.recompile()
+
+
+def run_placeholder_naming_pass(
+    gm: torch.fx.GraphModule,
+    export_graph_signature: torch.export.ExportGraphSignature,
+    constants: Dict[str, Union[torch.Tensor, torch._C.ScriptObject]],
+    mod: torch.nn.Module,
+    fake_args,
+    fake_kwargs,
+    fake_params_buffers,
+) -> None:
+    """
+    At the end of _export_non_strict(), run this pass to assign better placeholder node names.
+    The default names are arg0_1, arg1_1, ..., which are not very informative.
+    The names assigned by this pass are:
+        - User inputs:
+            These follow the signature of mod.forward(), e.g. forward(x, y) gets you x, y nodes.
+            If the tensors are nested in dictionaries/lists/tuples,
+            the names are a concatenation of the path to the tensor.
+                e.g. x = {
+                    'a': torch.randn(),
+                    'b': [torch.randn(), torch.randn()]
+                }
+            gets you nodes x_a, x_b_0, x_b_1.
```

Comment:
what happens if users just have *args as inputs? something like https://fburl.com/code/worpg00x

Reply:
Ah, this becomes `arg_0, arg_1` as tuple indexing for each argument, the forward() signature is used to bind inputs and name them.